### PR TITLE
docs: replacing documentation platform from sphinx to mkdocs.

### LINKED ISSUE #442: Better documentation site
### What did you had in mind?

Replace `sphinx` with `mkdocs` which seems better for this project.

For refs, [this](https://aioswitcher.tomfi.info/) is the docs site for this project build with `sphinx`,
and [this](https://archetypes.tomfi.info/) is a docs site built for another project of mine, build with `mkdocs`.

### Are you trying to fix a problem?

_No response_

### Any lead on how this feature can be implemented?

_No response_

### PR DESCRIPTION
**Pull request by dolby360**
remove all old docs and write new ones based on the previous ones.
setting new commands in toml file.
use some cool features to make the docs look cool.

Signed-off-by: dolby360 <mrdolbyking@gmail.com>

**Related issue (if any):** fixes #*442*

## Checklist

- [v] I have followed this repository's contributing guidelines.
- [v] I have used the conventional commits specification for my commit messages.
- [v] I have signed my commits.
- [v] I will adhere to the project's code of conduct.

### CODE REVIEW

**Comment by TomerFi** [comment]
@dolby360 Thank you very much.
Can you please rebase from release/v3 and resolve the conflicts before pushing again?

**Comment by dolby360** [comment]
> @dolby360 Thank you very much. Can you please rebase from release/v3 and resolve the conflicts before pushing again?

Ummm. Actually, I'm not sure what is going on here. seems like I'm rebased on release/v3 and there are no actual conflicts. docs/spelling_wordlist.txt is deleted and I added little something to scripts/discover_devices.py so...

**Comment by TomerFi** [comment]
these are probably modifications committed to the release/v3 branch after you forked it.
you just gonna have to rebase and resolve the conflicts.

I can do it this weekend if it'll still be relevant.

**Comment by dolby360** [comment]
> these are probably modifications committed to the release/v3 branch after you forked it. you just gonna have to rebase and resolve the conflicts.
> 
> I can do it this weekend if it'll still be relevant.

Please don't work on this I'll solve it. Thanks for the guidance.

**Comment by dolby360** [comment]
@TomerFi OK it seems like I solve the conflicts.

File: pyproject.toml [hunk 1]
```diff
@@ -32,22 +32,14 @@ exclude = [
...
-    optional = true
-
-    [tool.poetry.dependencies.sphinxcontrib-autoprogram]
-    version = "0.1.7"
-    optional = true
-
-    [tool.poetry.dependencies.sphinxcontrib-spelling]
-    version = "7.2.1"
-    optional = true
+  mkdocs = {extras = ["docs"], version = "^1.3.0"}
```
Comment (TomerFi) [inline pyproject.toml:35]: Docs dependencies related to the "docs" extras group, should have an `optional = true` property so they will not be installed outside the context of the "docs" group.
```suggestion
  mkdocs = {extras = ["docs"], version = "^1.3.0", optional = true}
```

File: pyproject.toml [hunk 2]
```diff
@@ -76,10 +68,6 @@ exclude = [
 
   [tool.poetry.extras]
   docs = [
```
Comment (TomerFi) [inline pyproject.toml:65]: The _docs_ extras group should contain the documentation's dependencies, given the dependencies will be set as optional, the only way to install them would be to include the _docs_ extras group.

File: pyproject.toml (hunk 1 above)
Comment (TomerFi) [inline pyproject.toml]: Docs dependencies related to the _docs_ extras group, should have an `optional = true` property so they will not be installed outside the context of the `docs` group.
```suggestion
  mkdocs = {extras = ["docs"], version = "^1.3.0", optional = true}
  ...
```

File: pyproject.toml [hunk 3]
```diff
@@ -76,10 +68,6 @@ exclude = [
...
-  "insegel",
-  "sphinx",
-  "sphinxcontrib-autoprogram",
-  "sphinxcontrib-spelling",
   "toml"
```
Comment (TomerFi) [inline pyproject.toml]: The _toml_ dependency was used in the `docs/conf.py` python script used to build the documentation with _sphinx_.
It used _toml_ to parse the `pyproject.toml` and grab the project name, version, etc.

Do we still need it?

File: scripts/discover_devices.py [hunk 4]
```diff
@@ -51,7 +51,8 @@
 Example output:
     Switcher devices broadcast a status message every approximately 4 seconds. This
     script listens for these messages and prints a serialized version of the to the
-    standard output, for example (note the ``device_id`` and ``mac_address`` properties)::
+    standard output, for example (note the device_id and mac_address properties):
```
Comment (TomerFi) [inline scripts/discover_devices.py]: I think the discover_devices script is being updated in another PR and this modification might cause conflicts.
Perhaps it's better to skip the modifications in this file as they are purely semantic.
Unless I'm missing something.

**Review by TomerFi — CHANGES_REQUESTED** [review]
Thank you very much for your contribution, it looks really good.

Besides the changes requested, I also have a couple of other points:

- The release workflow has a step called [Set new project version](https://github.com/dolby360/aioswitcher/blob/dolev/better_docs/.github/workflows/release.yml#L79), this step updates the auto-bumped semantic version in _pyproject.toml` before deploying, the next step, _Commit, tag, and push_ pushed the update file to the repository. We should include an update to the version in _mkdocks.yml_ at this point as well. You can see an example of using _sed_ for this in [another project of mine](https://github.com/TomerFi/tomfi-archetypes/blob/main/.github/workflows/release.yml#L97).
- I'm not able to get this working on my _Fedora_ using _Pyhton 3.10.5_ and _poetry 1.1.14_, it fails to install _editdistpy_, I'll look into this later or tomorrow, what environment are you working with?

**Comment by dolby360** [comment]
> Thank you very much for your contribution, it looks really good.
> 
> Besides the changes requested, I also have a couple of other points:
> 
> * The release workflow has a step called [Set new project version](https://github.com/dolby360/aioswitcher/blob/dolev/better_docs/.github/workflows/release.yml#L79), this step updates the auto-bumped semantic version in _pyproject.toml` before deploying, the next step, _Commit, tag, and push_ pushed the update file to the repository. We should include an update to the version in _mkdocks.yml_ at this point as well. You can see an example of using _sed_ for this in [another project of mine](https://github.com/TomerFi/tomfi-archetypes/blob/main/.github/workflows/release.yml#L97).
> * I'm not able to get this working on my _Fedora_ using _Pyhton 3.10.5_ and _poetry 1.1.14_, it fails to install _editdistpy_, I'll look into this later or tomorrow, what environment are you working with?

* OK I changed it.
* Win 11, Python 3.10.5, Poetry version 1.1.13.
How do you install that?  I want to try to reproduce it.

**Comment by TomerFi** [comment]
You'll need Linux os running to reproduce correctly, instead of that, maybe make this [workflow](https://github.com/TomerFi/aioswitcher/runs/7373927084?check_suite_focus=true) pass, I assume the one running after it passes should reproduce this issue.

**Comment by dolby360** [comment]
> You'll need Linux os running to reproduce correctly, instead of that, maybe make this [workflow](https://github.com/TomerFi/aioswitcher/runs/7373927084?check_suite_focus=true) pass, I assume the one running after it passes should reproduce this issue.

OK no problem, I'm working on making the build pass (While trying to **not** requesting many rebuilds).
BTW I mentioned you in a comment there, I think we should suppress line-length rule in some files for reviewdog to keep help docs as closest to the original output as possible. Please let me know if you agree.

P.S. 
thank you very much for being patient in this PR.

**Comment by TomerFi** [comment]
> OK no problem, I'm working on making the build pass (While trying to **not** requesting many rebuilds).
> BTW I mentioned you in a comment there, I think we should suppress line-length rule in some files for reviewdog to keep help docs as closest to the original output as possible. Please let me know if you agree.

Missed it - sure - got for it.

**Comment by dolby360** [comment]
> # [Codecov](https://codecov.io/gh/TomerFi/aioswitcher/pull/515?src=pr&el=h1&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=Tomer+Figenblat) Report
> > ❗ No coverage uploaded for pull request base (`release/v3@a550567`). [Click here to learn what that means](https://docs.codecov.io/docs/error-reference?utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=Tomer+Figenblat#section-missing-base-commit).
> > The diff coverage is `n/a`.
> 
> ```diff
> @@              Coverage Diff               @@
> ##             release/v3      #515   +/-   ##
> ==============================================
>   Coverage              ?   100.00%           
> ==============================================
>   Files                 ?        10           
>   Lines                 ?      1037           
>   Branches              ?         0           
> ==============================================
>   Hits                  ?      1037           
>   Misses                ?         0           
>   Partials              ?         0           
> ```

@TomerFi is this the problem you mentioned?

**Comment by dolby360** [comment]
> # [Codecov](https://codecov.io/gh/TomerFi/aioswitcher/pull/515?src=pr&el=h1&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=Tomer+Figenblat) Report
> > ❗ No coverage uploaded for pull request base (`release/v3@a550567`). [Click here to learn what that means](https://docs.codecov.io/docs/error-reference?utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=Tomer+Figenblat#section-missing-base-commit).
> > The diff coverage is `n/a`.
> 
> ```diff
> @@              Coverage Diff               @@
> ##             release/v3      #515   +/-   ##
> ==============================================
>   Coverage              ?   100.00%           
> ==============================================
>   Files                 ?        10           
>   Lines                 ?      1037           
>   Branches              ?         0           
> ==============================================
>   Hits                  ?      1037           
>   Misses                ?         0           
>   Partials              ?         0           
> ```

@TomerFi OK it seems like this is happening in the other PRs as well.
Regarding your issue, please give me more details I have WSL2 (also docker) I can reproduce whatever you want.

**Comment by TomerFi** [comment]
You can ignore the coverage report reported in the comments, it looks like it's not set to track the `release/v3` branch, I'll take a look at it, but I don't think we should be concerned about it.

As for the merging, the site won't build on my PC.
Maybe it's an issue on my side, I'll take a look at it this weekend to verify before merging this.

**Comment by TomerFi** [comment]
Sorry for the delay, I was preoccupied, will take a look at this tonight, just want to verify it works on my end and I will merge this - thank you very much for your contributions.

**Comment by dolby360** [comment]
> Sorry for the delay, I was preoccupied, will take a look at this tonight, just want to verify it works on my end and I will merge this - thank you very much for your contributions.

It's ok take your time. If you need me to add something else to the pipeline or anything, just ask.

**Comment by TomerFi** [comment]
@dolby360 
Please check out my above-added commits, I fixed a couple of minor stuff and pushed them directly to your fork,
please update your local repo.
The most important thing I added, is a job in the CI workflow that lints and builds the documentation site.

It passes, to my surprise (more on that later), but it raises two issues for me:

---

take a look at this [workflow run](https://github.com/TomerFi/aioswitcher/runs/7507941505?check_suite_focus=true),
more specifically the output of the `Build documentation site` step in the `Verify documentation site` job:

image

Is there any way bumping these bad-boys to `Errors`, I'm afraid we might miss them as `Warnings` and lose sync to the code docs over time?

----

And now for the second, more intriguing issue, which also relates to a comment I made earlier: https://github.com/TomerFi/aioswitcher/pull/515#issuecomment-1185511527
the reason I'm surprised the new workflow passed is that... it fails on my PC:

image

**...**

image

Based on my intregation so far, the cause for this is, obviously the `editdistpy` module, which is a transitive dependency of `symspellpy`, which is a transitive dependency of `mkdocs-spellcheck` which you added in this PR.

As evident in my error log, and as can be verified in `editdistpy`'s [Built Distributions](https://pypi.org/project/editdistpy/#files), regardless of the OS, there's no wheel for Python 3.10. meaning that while building, Python downloads the sources and builds the module.

So I assumed your Windows 11 OS has a couple of C extensions in place, missing from my Linux-based OS and required for the module build.
But then the new workflow job passed, which means that GitHub's Ubuntu runner also has the required C extensions missing from my Fedora.

So I guess I just find out what OS package I'm missing to be able to build `editdistpy`...

**Comment by TomerFi** [comment]
@dolby360 
Fixed my issue, I was missing the `python3-devel` package which holds the missing header file, `Python.h` used [here](https://github.com/mammothb/editdistpy/blob/c7f71bc521b793fc61d5b02e9caeae5546fceef3/editdistpy/levenshtein.cpp#L27).

```
sudo dnf install python3-devel
```

I can now build the site on my PC it looks amazing - thank you very much!
I just have a couple of small notes, and I will post them back here today.
Thank you for your patience.

**Review by TomerFi — APPROVED** [review]
@dolby360 this is very good work!
Thank you very much - merging now!

**Comment by dolby360** [comment]
@TomerFi I'm on vacation without my computer. I hope I can check everything out this weekend.

**Comment by dolby360** [comment]
Yayyyy

**Pull request merged**